darts/explainability/shap_explainer.py

```diff
@@ -0,0 +1,474 @@
+
+        # We suppose for now the output_chunk_length existence
+        raise_if(
+            max(horizons)>self.model.output_chunk_length-1,
+            "One of the horizons is greater than the model output_chunk_length."
+        )
+
+        if nb_samples:
+            foreground_X_sampled = shap.utils.sample(self.explainers.background_X, nb_samples)
+        else:
+            foreground_X_sampled = self.explainers.background_X
+
+        shap_values = []
+        if target_names is None:
+            target_names = self.target_names
+        if horizons is None:
+            horizons = range(self.model.output_chunk_length)
+
+        for t in target_names:
+            for h in horizons:
+                shap_values.append(self.explainers.shap_values_from_X(foreground_X_sampled ,h ,t))
+                plt.title('Target: `{}` - Horizon: {}'.format(t, 't+' + str(h)))
+                shap.summary_plot(shap_values[-1], foreground_X_sampled,  plot_type=plot_type)
+
+class RegressionShapExplainers:
+    """
+    Helper Class to wrap the different cases we encounter with shap different explainers, multivariates, 
+    horizon etc.
+    Aim to provide shap values for any type of RegressionModel. Manage the MultioutputRegressor cases.
+    For darts RegressionModel only.
+    TODO implement a test to not recompute each time the shap values in case of multioutput flag is True.
+    """
+    
+    def __init__(
+        self, 
+        model: GlobalForecastingModel,
+        background_series: Union[TimeSeries, Sequence[TimeSeries]],
+        background_past_covariates: Union[TimeSeries, Sequence[TimeSeries]],
+        background_future_covariates: Union[TimeSeries, Sequence[TimeSeries]],
+        background_nb_samples: Optional[int] = None
+    ):
+
+        self.model = model
+        self.multioutput = self.model.model._get_tags()['multioutput']
+        self.target_dim =self.model.input_dim['target']
+
+        self.background_X = self._create_RegressionModel_shap_X(
+            background_series,
+            background_past_covariates,
+            background_future_covariates,
+            background_nb_samples
+        )
+
+        self.target_dict = {c:i for i, c in enumerate(background_series.columns)}
+
+        if (self.target_dim > 1 or self.model.output_chunk_length > 1) and (not self.multioutput):
+            self.explainers = {}
+            for i in range(self.model.output_chunk_length):
+                self.explainers[i] = {}
+                for j in range(self.target_dim):
+                    self.explainers[i][j] = shap.Explainer(
+                        self.model.model.estimators_[i+j], 
+                        self.background_X
+                    )
+                    # Special case of trees, where we don't want to depend on a background (faster in general)
+                    # Warning: Randomforest is quite slow.
+                    if isinstance(self.explainers[i][j], shap.explainers._tree.Tree):
```

Comment:
Isn't there another way to check this than accessing a private member?

Reply:
It seems to me that here we are first building an explainer, then checking its type, and then changing it based on the result. Shouldn't we build the right type of explainer right away? E.g. by checking whether the initial estimator object is a tree?